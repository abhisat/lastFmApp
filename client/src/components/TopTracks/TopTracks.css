.cardDescription {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}
.photoCard {
  margin: 0 auto !important;
  margin-top: 20px !important;
}
.backButton {
  margin: 10px !important;
  background-color: transparent !important;
  color: #b90000 !important;
  font-size: 2rem !important;
  padding: 1rem !important;
}

.photoCardContainer {
  width: 100%;
}
