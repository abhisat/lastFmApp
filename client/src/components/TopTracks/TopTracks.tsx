import React, { useState, useEffect, SyntheticEvent, ReactText } from "react";
import axios from "axios";
import { SearchBar } from "../SearchBar/SearchBar";
import { ResultList } from "../ResultList/ResultList";
import Logo from "../../assets/logo.png";
import {
  Image,
  Container,
  PaginationProps,
  Button,
  Icon
} from "semantic-ui-react";
import { TrackResponse, ResponseItem } from "../../types/ResponseTypes";
import Loader from "react-loader-spinner";
import { RouteComponentProps } from "react-router-dom";
import { Pagination } from "semantic-ui-react";
import "./TopTracks.css";

const TopTracks: React.FunctionComponent<RouteComponentProps> = props => {
  const [feedTitle, setFeedTitle] = useState<string>();
  const [feeds, setFeeds] = useState<ResponseItem[]>();
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [activePage, setActivePage] = useState<ReactText>(1);
  const [totalPages, setTotalPages] = useState<number>();
  const itemsPerPage = "5";
  const { history } = props;

  useEffect(() => {
    fetchFeeds();
  }, [activePage]);

  const fetchFeeds = async function() {
    const result = await axios(getSearchURL());
    const parsedFeeds: TrackResponse = result.data.toptracks;
    setTotalPages(parsedFeeds["@attr"]["totalPages"]);
    parsedFeeds.track.length > 5
      ? setFeeds(parsedFeeds.track.slice(parsedFeeds.track.length - 5))
      : setFeeds(parsedFeeds.track);

    setIsLoading(false);
  };

  const getSearchURL = () => {
    const corsServerURL = process.env.REACT_APP_CORS_SERVER;
    let baseAPIUrl = new URL(process.env.REACT_APP_BASE_URL);
    const urlParam = new URLSearchParams(props.location.search);
    const artist = urlParam.has("q") ? urlParam.get("q") : null;
    setFeedTitle(`Top Tracks by ${artist}`);
    baseAPIUrl.searchParams.set("artist", artist);
    baseAPIUrl.searchParams.set("method", "artist.gettoptracks");
    baseAPIUrl.searchParams.set("api_key", "9bfb0463ecfb5dd130cb40efbd898af0");
    baseAPIUrl.searchParams.set("format", "json");
    baseAPIUrl.searchParams.set("limit", itemsPerPage);
    baseAPIUrl.searchParams.set("page", activePage.toString());
    return corsServerURL + baseAPIUrl;
  };

  const handlePaginationClick = (
    e: SyntheticEvent,
    pageInfo: PaginationProps
  ) => {
    let pageNumber: ReactText = pageInfo.activePage;
    setIsLoading(true);
    setActivePage(pageNumber);
  };

  const handleBackButtonClick = () => {
    props.history.goBack();
  };

  return (
    <React.Fragment>
      <Container className='appContainer'>
        <Image src={Logo} className={"logo"}></Image>
        <SearchBar />
        <Button
          className='ui button backButton'
          onClick={handleBackButtonClick}
        >
          <Icon name='arrow left' />
        </Button>
        {isLoading ? (
          <Loader type='TailSpin' color={"#b90000"} height={100} width={100} />
        ) : (
          <ResultList title={feedTitle} feeds={feeds} isLoading={isLoading} />
        )}
        <Pagination
          className={"pagination"}
          activePage={activePage}
          onPageChange={handlePaginationClick}
          totalPages={totalPages}
          ellipsisItem={null}
        />
      </Container>
    </React.Fragment>
  );
};

export { TopTracks };
