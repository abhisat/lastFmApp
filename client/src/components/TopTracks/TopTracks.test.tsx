import React from "react";
import { TopTracks } from "./TopTracks";
import EnzymeAdapter from "enzyme-adapter-react-16";
import Enzyme, { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { createMemoryHistory, createLocation } from "history";
import { match } from "react-router";

Enzyme.configure({
  adapter: new EnzymeAdapter()
});

const handlebackButtonClick: jest.Mock = jest.fn();

const mountApp = () => {
  const history = createMemoryHistory();
  const path = `/route/:id`;
  const match: match<{ id: string }> = {
    isExact: false,
    path,
    url: path.replace(":id", "1"),
    params: { id: "1" }
  };
  const location = createLocation(match.url);
  const wrapper: Enzyme.ShallowWrapper = shallow(
    <TopTracks history={history} location={location} match={match} />
  );
  wrapper.update();
  return wrapper;
};

describe("<TopTracks/>", () => {
  let App: Enzyme.ShallowWrapper;

  beforeEach(() => {
    App = mountApp();
  });

  afterEach(() => {});

  it("renders without crashing", () => {
    mountApp();
  });

  it("Matches the snapshot", () => {
    expect(toJson(App)).toMatchSnapshot();
  });
});
