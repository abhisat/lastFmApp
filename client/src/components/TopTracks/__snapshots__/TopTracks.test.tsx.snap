// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopTracks/> Matches the snapshot 1`] = `
<Fragment>
  <Container
    className="appContainer"
  >
    <Image
      as="img"
      className="logo"
      src="logo.png"
      ui={true}
    />
    <withRouter() />
    <Button
      as="button"
      className="ui button backButton"
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="arrow left"
      />
    </Button>
    <Loader
      className=""
      color="#b90000"
      height={100}
      style={Object {}}
      timeout={0}
      type="TailSpin"
      visible={true}
      width={100}
    />
    <Pagination
      activePage={1}
      aria-label="Pagination Navigation"
      boundaryRange={1}
      className="pagination"
      ellipsisItem={null}
      firstItem={
        Object {
          "aria-label": "First item",
          "content": "«",
        }
      }
      lastItem={
        Object {
          "aria-label": "Last item",
          "content": "»",
        }
      }
      nextItem={
        Object {
          "aria-label": "Next item",
          "content": "⟩",
        }
      }
      onPageChange={[Function]}
      pageItem={Object {}}
      prevItem={
        Object {
          "aria-label": "Previous item",
          "content": "⟨",
        }
      }
      siblingRange={1}
    />
  </Container>
</Fragment>
`;
